### PR TITLE
Axial Diffusion

### LINKED ISSUE #1651: Support for axial diffusion
Cf. "Mechanism of signal spreading" in "Biophysics of Biochemical Signaling in Dendritic Spines: Implications in Synaptic Plasticity" by Yasuda (2017) https://doi.org/10.1016/j.bpj.2017.07.029.

"Axial" is along the cable.

### PR DESCRIPTION
**Pull request by thorstenhater**
# Introduction

Plasticity processes are mediated by signalling ions, eg Ca++, which are generated by 
synapses upon reception of a spike. These ions then propagate along the dendrite 
by diffusion and are received by other synapses, modifying their weight upon reception.

We introduce a simplified -- as opposed to solving the Nernst-Planck equation -- approach
to model these processes. Concretely, we will introduce a new field, dubbed `Xd`, to the 
NMODL dynamics, with the following properties
- `Xd` can be read from and written to by NMODL density and point mechanisms.
   This allows us to model decay, consumption, and sources, as well as interaction with the actual processes.
- `Xd` propagates according to a diffusion law `∂_t Xd = ∂_z 1/rX ∂_z Xd + iX/qi` along the neuron, where
   the per-species resistivity `rX` may vary in space (accessible via the `paint`-interface) but not in time.
- **NOTE**: Currently this follows a simplified update model, where `Xd` is atomically updated during the
   `advance_state` phase and **not** like other ionic quantities in two steps (once as a `STATE` and later effects
     are applied in `write_ions` / `compute_currents`.)
    - We could get away with this as there is no precedent in NRN and we can make our own semantics to fit our needs
    - However diverting from the ionic model feels gross.

## Update Nov 2021

After the initial exploration and design a number of issues became apparent.

1. Arbor (and Neuron) model the neuron as a thin cable in the following limit case
  - External and internal concentrations are constant in time; as they are linked to an infinite buffer of the relevant species.
  - Currents `iX` across the membrane do not contribute to `Xi`/`Xo`.
  - Similarly, axial currents to not change `Xi`.
  - Neither are ions transported by concentration gradients.
  - Therefore, `Xi` and `Xo` are reset every `dt` and we cannot use `Xi` for our model.
2. Partial presence of a diffusing species is non-sensical
  - Basically, the matrix solver requires a defined value everywhere 
3. All dynamics are only well-defined in proximity of the membrane, essentially ignoring the interior of the neuron. 
4. Concentration models write to a state variable shadowing `Xi`

# Changes

- Add `paint` and `default` interfaces.
  - Diffusivity is `paint`able on regions like `init_ext_conc`
  - ... and can be added as a default via `set_ion` (Python) or `set_default` (C++)
  - see `python/example/diffusion.py`

- Add per-ion conductivity `gX`, needed for matrix solver
  - part of the ABI, got bumped to 0.1.1
  - modcc computes and accumulates `gX` on `compute_currents` iff `iX` is written
  - `shared_state` sets up the backing store

- Add per-ion diffusion solvers

- Clean-up
  - Remove redundant arrays
  - Remove matrix.hpp
  - Push matrix_state into shared_state

# Open Questions/Issues

- (How to) disable ion-conductivity if we do not need it?
- Currently the cable equation and diffusion solvers (1 + 1/diffusing ion) store duplicates of the solver state (eg parent indices), it would be nice to have one copy in total.
- What to do if diffusivity defaults to a finite value, but is overridden to be zero in a region?

# References

# Linked Issues

Closes #1651

### CODE REVIEW

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
Finally this is in shape. @schmitts please double check if it does suit your needs.

**Comment by thorstenhater** [comment]
bors try

**Comment by brenthuisman** [comment]
bors try

**Comment by thorstenhater** [comment]
Again an Ault fail state, Mr Bors

**Review by AdhocMan — CHANGES_REQUESTED** [review]
I've only found a few small issues

File: arbor/fvm_lowered_cell_impl.hpp [hunk 1]
```diff
@@ -501,11 +497,26 @@ fvm_initialization_data fvm_lowered_cell_impl<Backend>::initialize(
                 D.init_membrane_potential, D.temperature_K, D.diam_um, std::move(src_to_spike),
                 data_alignment? data_alignment: 1u);
 
+    state_->solver =
+        {D.geometry.cv_parent, D.geometry.cell_cv_divs, D.cv_capacitance, D.face_conductance, D.cv_area, fvm_info.cell_to_intdom};
+
     // Instantiate mechanisms, ions, and stimuli.
+    auto mk_diff_solver = [&](const auto& fd) {
+        // TODO(TH) _all_ solvers should share their RO data, ie everything below D here.
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/fvm_lowered_cell_impl.hpp:505]: No, this is a longer term marker for me (and for others, if someone likes to pick it up). I add these
comments for things that would be nice to have, but do not warrant an issue.

This particular item requires some thinking on safety and lifetimes. For now it just works the way it
is, the suggestion is a potential optimisation.

File: arbor/include/arbor/cable_cell_param.hpp [hunk 2]
```diff
@@ -32,6 +32,7 @@ struct cable_cell_ion_data {
     std::optional<double> init_int_concentration;
     std::optional<double> init_ext_concentration;
     std::optional<double> init_reversal_potential;
+    std::optional<double> diffusivity = {0.0};
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/include/arbor/cable_cell_param.hpp:35]: That's an oversight, thanks!

File: example/diffusion/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# 'single' example.
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply example/diffusion/README.md]: Made a new Readme ;)

**Review by AdhocMan — APPROVED** [review]
Looks good from my side.

**Pull request merged**